src/VisualStudio/Core/Def/Implementation/Venus/ContainedDocument.DocumentServiceProvider.cs

```diff
@@ -0,0 +1,278 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Projection;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Venus
+{
+    internal sealed partial class ContainedDocument
+    {
+        // this is to support old venus/razor case. all new razor should use thier own implementation not ours
+        public class DocumentServiceProvider : IDocumentServiceProvider
+        {
+            public static readonly DocumentServiceProvider Instace = new DocumentServiceProvider();
+
+            private readonly SpanMapper _spanMapper;
+            private readonly DocumentExcerpter _excerpter;
+
+            private DocumentServiceProvider()
+            {
+                _spanMapper = new SpanMapper();
+                _excerpter = new DocumentExcerpter();
+            }
+
+            public TService GetService<TService>() where TService : class, IDocumentService
+            {
+                if (_spanMapper is TService spanMapper)
+                {
+                    return spanMapper;
+                }
+
+                if (_excerpter is TService excerpter)
+                {
+                    return excerpter;
+                }
+
+                // ask the default document service provider
+                return TextDocumentState.DefaultDocumentServiceProvider.Instance.GetService<TService>();
+            }
+
+            private static IProjectionSnapshot GetProjectSnapshot(SourceText sourceText)
+            {
+                return sourceText.FindCorrespondingEditorTextSnapshot() as IProjectionSnapshot;
+            }
+
+            private class SpanMapper : ISpanMappingService
+            {
+                public async Task<ImmutableArray<MappedSpanResult>> MapSpansAsync(Document document, IEnumerable<TextSpan> spans, CancellationToken cancellationToken)
```

Comment:
takes in an IENumerable, but returns an ImmutableArray?  Can we make these match (ideally ImmutableArray)